tcl/org.testeditor.tcl.dsl.ide/src/org/testeditor/tcl/dsl/ide/TclContentProposalProvider.xtend

```diff
@@ -0,0 +1,296 @@
+		val macros = macroCollection.macros
+		macros.map[template].makeTemplateProposals(context, acceptor)
+	}
+
+	private def makeTemplateProposals(Iterable<Template> templates, ContentAssistContext context, IIdeContentProposalAcceptor acceptor) {
+		val mandatoryPrefix = context.mandatoryPrefixForTemplate(true)
+		templates.forEach [
+			makeTemplateProposal(mandatoryPrefix, context, acceptor)
+		]
+	}
+
+	private def makeTemplateProposal(Template template, String mandatoryPrefix, ContentAssistContext context, IIdeContentProposalAcceptor acceptor) {
+		val element = template.restoreString
+		if (element.startsWith(mandatoryPrefix)) {
+			val actualElement = context.prefix + element.substring(mandatoryPrefix.length).trim
+			val proposal = proposalCreator.createProposal(actualElement, context)
+			if (proposal !== null) {
+				acceptor.accept(proposal => [
+					label = element
+					description = 'STEP'
+				], proposalPriorities.getCrossRefPriority(null, proposal))
+			}
+		}
+	}
+
+	private def makeElementParamsProposal(StepContentElement element, ContentAssistContext context, IIdeContentProposalAcceptor acceptor) {
+		val model = EcoreUtil2.getContainerOfType(element, TestStep)
+		model.makeElementParamsProposal(element.value, context, acceptor)
+	}
+
+	/**
+	 * return elements that are available for the given interaction and have the given prefix
+	 */
+	private def Iterable<ComponentElement> filterElementsApplicable(Iterable<ComponentElement> elements, InteractionType interaction, String prefix) {
+		return elements.filter [
+			val interactionTypes = componentElementInteractionTypes
+			val validCandidate = interactionTypes.contains(interaction) && name.startsWith(prefix)
+			if (logger.traceEnabled) {
+				logger.trace('''Element-Parameter validForProposal='«validCandidate»' name='«name»'.''')
+			}
+			return validCandidate
+		]
+	}
+
+	private def makeElementParamsProposal(TestStep model, String prefix, ContentAssistContext context, IIdeContentProposalAcceptor acceptor) {
+		if (!context.prefix.isNullOrEmpty) {
+			val componentContext = EcoreUtil2.getContainerOfType(model, ComponentTestStepContext)
+			val component = componentContext.component
+			val interaction = model.interaction
+			val applicableElements = component.elements.filterElementsApplicable(interaction, prefix)
+			val currentNode = context.currentNode
+			val includeClosingBracket = !currentNode.text.contains('>') && (currentNode.nextSibling === null || !currentNode.nextSibling.text.contains('>'))
+			applicableElements.forEach [ el |
+				val entry = '''«IF prefix.isNullOrEmpty»<«ENDIF»«el.name»«IF includeClosingBracket»>«ENDIF»'''
+				if (logger.traceEnabled) {
+					logger.trace('''Element-Parameter-Proposal name='«el.name»', context.prefix='«context.prefix»', proposal entry='«entry»'.''')
+				}
+				val proposal = proposalCreator.createProposal(entry, context)
+				if (proposal !== null) {
```

Comment:
Is this an opportunity for method extraction? A similar pattern (`if (proposal !== null) { acceptor.accept(…`) appears a couple of times, I think

Reply:
the if is more of a guard for the following access on inner variables via proposal => [  ... ]